Rewrite

iliis opened the pull request:
Hoi Jörn, hallo Pascal.
Befor ich komplett vom Stuhl falle hier mal was ich heut noch an Arbeit reinquetschen konnte. Ich hätt von Anfang an so coden sollen. Naja, jetzt ists eben so.
Mit der Deadline seh ichs grad echt nicht :( Ja, das geht auf meine Kappe...
etwas übermüdete Grüsse,
Sam

iliis committed:
Changed some details after last discussion.

iliis committed:
Cleanup. Add sceletons for core functions.

iliis committed:
Implement core functions.

Basic warping seems to be working.

TODO:
- derive/implement Jacobis
- write more tests (especially hand picked samples) for:
    - camera_transformation
    - camera_project
- write helper function to read in complete test scene (with ground
  truth etc.)
- clean up helper functions (remove unnecessary ones, document and test others)

iliis committed:
Add helpful diagram to show coordinate system orientations.

rehderj commented on matlab/core/camera_project.m:
```diff
@@ -0,0 +1,42 @@
+% Z-Axis points 'outwards' (towards viewer)
+% => points in front of camera have negative Z values and thence comes the
+%    minus in front of points_world(:,3)
+points_camera = points_world(:,1:2) ./ repmat(-points_world(:,3),1,2) .* intrinsics.focal_length + repmat(intrinsics.principal_point,N,1);
```
Woher das minus: `-points_world(:,3)`?

rehderj commented on matlab/core/camera_project_inverse.m:
```diff
@@ -0,0 +1,45 @@
+% project into world coordinates
+points_world = [ ...
+    (points_camera - repmat(intrinsics.principal_point,N,1)) .* repmat(depths,1,2) ./ intrinsics.focal_length ...
+    -depths ]; % our Z-Axis points 'outwards' (toward viewer) => positive depth = negative Z
```
`-depths` ?

iliis replied:
Für das und das vorherige (`-points_world(:,3)`) siehe Kommentare, bzw. die Definition der Koordinatensysteme in https://github.com/iliis/BA/blob/rewrite/matlab/coordinate_systems.pdf
Die Z-Achse zeigt bei mir weg von der Szene:
Z <----- Kamera .<
Das ist bewusst gewählt, weil ich denk die Wahl der Achsen spielt nicht soo ne Rolle (Mal ein Minus hier vs. 1:1 Kompatibilität mit Blender).
Edit: Und ich gehe davon aus, dass ich ein Tiefenbild mit positiven Werten krieg, also eigentlich gespiegelt. Das ist tatsächlich nicht so elegant...

rehderj replied:
Ich mag, dass Du eine Dokument angelegt hast, das die Konventionen visualisiert. Die Konvention selbst mag ich nicht und empfehle eher Kompatibilität mit der Computer Vision-Literatur als mit Blender.

iliis replied:
Die gibt es, die Konvention? ^^ Ich hatte bisher den Eindruck, dass jeder ein bisschen was anderes macht. Zumindest in der Computer Graphics-Welt...
Hast du ne Referenz? [citation needed] sozusagen, dann übernehm ich das schon. Solangs wirklich klar definiert ist, ists mir wayne.

rehderj replied:
Eigentlich wird das schon sehr einheitlich gehandhabt, ein Dokument dazu findest Du zum Beispiel [hier](http://www.igp.ethz.ch/photogrammetry/education/lehrveranstaltungen/PhotogrammetryHS12/04-E-camera.pdf).

iliis replied:
Ja, das ist gut dokumentiert. Aber da wirst du auch einige Minuszeichen kriegen. Die X-Achse der Bildebene (Folie 10) ist zum Beispiel invertiert zur globalen X-Achse (Folie 6, wobei dort die Bild-X-Achse wiederum parallel zur globalen eingezeichnet ist).
Und die Bildkoordinaten beginnen unten links (in Matlab wärs oben links per default).

rehderj commented on matlab/core/camera_transform.m:
```diff
@@ -0,0 +1,47 @@
+assert(all(size(T_keyframe_to_current) == [1 6]));
+
+% apply transformation
+points_current = points_keyframe * angle2dcm(T_keyframe_to_current(4:6)) + repmat(T_keyframe_to_current(1:3), N, 1);
```
Ich mag das Konventionelle:  `R*X+T`

iliis replied:
Ah, das ist vertauscht so? Das wird geändert. Und die Funktion hat eh auch noch keine Tests.

rehderj replied:
Naja, es ist transponiert. In der Literatur ist die andere Form aber gebräuchlicher und ich empfehle sie entsprechend :)

iliis committed:
Fix camera_transform: Use conventional R*X+T.

iliis committed:
Update coordinate system chart to reflect JR's wishes.

WARNING: code was NOT updated! Chart and code do NOT correspond!

iliis committed:
Use column vectors instead of row vectors.

Add tests for image_to_list/list_to_image.
Add script to run all testscripts.

iliis committed:
Update intensity_sample() to use new coordinate systems.

[1,1] = bottom left
[W,H] = top right

iliis committed:
Switched to new coordinate system (see coordinate_systems.pdf)

TODO: write [more] tests for all core functions

iliis committed:
Add sceletons for transformation tests and utilities.

TODO: Fix Euler angles. Fix direction of transformation (move Camera or
move points).

iliis committed:
Implemented Jacobians. No tests, apart from intensity_sample().

iliis committed:
Implement camera_warp(): combines all the core functions.

Unfortunately, it's not very clean (flipped V-axis, no tests)

iliis committed:
Play around with Jacobians.

Not much luck: Real slow and not very correct :(
